IPython/utils/tokenutil.py

```diff
@@ -2,20 +2,29 @@
 
 # Copyright (c) IPython Development Team.
 # Distributed under the terms of the Modified BSD License.
+from __future__ import annotations
 
-from collections import namedtuple
+import tokenize
 from io import StringIO
 from keyword import iskeyword
-
-import tokenize
 from tokenize import TokenInfo
-from typing import List, Optional
+from typing import Generator, NamedTuple
 
 
-Token = namedtuple('Token', ['token', 'text', 'start', 'end', 'line'])
+class Token(NamedTuple):
+    token: int
+    text: str
+    start: int
+    end: int
+    line: str
 
-def generate_tokens(readline):
+
+def generate_tokens(text_or_readline) -> Generator[TokenInfo, None, None]:
     """wrap generate_tkens to catch EOF errors"""
+    if isinstance(text_or_readline, str):
+        readline = StringIO(text_or_readline).readline
+    else:
+        readline = text_or_readline
```

Comment:
I have a few hours in a train so I'll do a cursory review. 

The comments bellow are not a request to change (at least for now), I'm laying down my thoughts, and will come back to it. 

I would prefer to not do that; I'm in general not a big fan of function that takes multiple types like this as their input; It's also meant to mimick tokenize.generate_token, which cannot take a str. 

And yes, I hate that it can't take str; I get annoyed at it everytime. 

Also _technically_ renaming the parameter is a breakage of API as technically a positional parameter can be passed as a named argument.

We could audit all places that uses generate_tokens and move to str only as well; I woudl not be against.

Reply:
I'll leave as the original